### PR TITLE
Coachmark: Migrate Coachmark and associated components to function components

### LINKED ISSUE #14340: Coachmark: strict mode compliance (UNSAFE_)
Coachmark and related components have the following strict mode issues:
- Coachmark: `componentWillReceiveProps`
- PositioningContainer (a simplified fork of Callout):
  - `componentWillMount`
  - `componentWillUpdate`

### PR DESCRIPTION
**Pull request by MLoughry**
#### Pull request checklist

- [x] Fixes #14340
- [x] Include a change request file using `$ yarn change`

#### Description of changes

This change migrates `Coachmark` and its associated components `Beak` and `PositioningContainer` to function components in the `react-next` package.

#### Focus areas to test

(optional)

### CODE REVIEW

**Comment by xugao** [comment]
I am concerned how we can better validate this change since there is almost no tests for coachmark

File: packages/react-next/etc/react-next.api.md [hunk 1]
```diff
@@ -893,6 +857,8 @@ export interface IPositioningContainerProps extends IBaseProps<IPositioningConta
     ariaLabelledBy?: string;
     backgroundColor?: string;
     bounds?: IRectangle;
+    // Warning: (ae-forgotten-export) The symbol "React" needs to be exported by the entry point index.d.ts
+    children?: React_2.ReactNode;
```
Comment (xugao) [inline packages/react-next/etc/react-next.api.md:861]: this looks unexpected

**Comment by MLoughry** [comment]
@xugao I've added a Screener test that needs approval

**Comment by JustSlone** [comment]
@xugao can you take a look at the added screener tests?

File: apps/vr-tests/src/stories/CoachmarkNext.stories.tsx [hunk 2]
```diff
@@ -0,0 +1,71 @@
...
+          ariaLabelledBy={'example-label1'}
+        >
+          Welcome to the land of Coachmarks!
+        </TeachingBubbleContent>
+      </Coachmark>
+    </>
+  );
+};
+
+storiesOf('Coachmark', module)
```
Comment (xugao) [inline apps/vr-tests/src/stories/CoachmarkNext.stories.tsx:50]:
```suggestion
storiesOf('Coachmark Next', module)
```

File: apps/vr-tests/src/stories/CoachmarkNext.stories.tsx [hunk 3]
```diff
@@ -0,0 +1,71 @@
+/*! Copyright (c) Microsoft Corporation. All rights reserved. Licensed under the MIT license. */
+import * as React from 'react';
```
Comment (xugao) [inline apps/vr-tests/src/stories/CoachmarkNext.stories.tsx:2]: could you duplicate the this test for Coachmark in OUFR as well? similar to other tests we have. so we always have 2 results side by side

**Comment by xugao** [comment]
@MLoughry - [this screener](https://screener.io/v2/states/Skgr32fUL.microsoft-fluentui/functional%2FCoachmark/1024x768/Chrome/coachmark-expanded-default-0) doesn't look correct to me

**Comment by MLoughry** [comment]
> @MLoughry - [this screener](https://screener.io/v2/states/Skgr32fUL.microsoft-fluentui/functional%2FCoachmark/1024x768/Chrome/coachmark-expanded-default-0) doesn't look correct to me

Yes, there's a logic bug in the original OUFR control when you provide `isCollapsed={false}` to start with. I fixed the bug when I migrated the control.

File: apps/vr-tests/src/stories/Coachmark.stories.tsx [hunk 4]
```diff
@@ -0,0 +1,71 @@
+/*! Copyright (c) Microsoft Corporation. All rights reserved. Licensed under the MIT license. */
```
Comment (ecraig12345) [inline apps/vr-tests/src/stories/Coachmark.stories.tsx:1]: Can you get rid of this header? (same in other files)

File: packages/react-next/src/components/Coachmark/Coachmark.types.ts [hunk 5]
```diff
@@ -229,21 +233,11 @@ export interface ICoachmarkStyleProps {
    */
   isCollapsed: boolean;
 
-  /**
```
Comment (ecraig12345) [inline packages/react-next/src/components/Coachmark/Coachmark.types.ts:232]: Why were these props removed? Even in a major release, we typically try not to remove props that weren't previously marked as deprecated. Though if there's a strong argument for removal you could go ahead with it and also mark the props as deprecated in OUFR to provide at least a bit of warning.

Also, if you go ahead with removing these, you should add a note about that (and any other breaking changes) to react-next's RELEASE_NOTES.md. Same for other PRs. (We just decided on this procedure yesterday, which is why you haven't heard about it before.)

File: packages/react-next/src/components/Coachmark/Beak/Beak.tsx [hunk 6]
```diff
@@ -8,78 +8,73 @@ import { RectangleEdge } from 'office-ui-fabric-react/lib/utilities/positioning'
...
-    super(props);
+export const Beak = React.forwardRef((props: IBeakProps, forwardedRef: React.Ref<HTMLDivElement>) => {
+  const { left, top, bottom, right, color, direction = RectangleEdge.top } = props;
 
-    initializeComponentRef(this);
-  }
-
-  public render(): JSX.Element {
-    const { left, top, bottom, right, color, direction = RectangleEdge.top } = this.props;
+  React.useImperativeHandle(props.componentRef, () => ({}), []);
```
Comment (ecraig12345) [inline packages/react-next/src/components/Coachmark/Beak/Beak.tsx:14]: Is this supposed to do anything? If not you should probably just remove the componentRef prop (and deprecate it in v7).

File: packages/react-next/src/components/Coachmark/Coachmark.styles.ts [hunk 7]
```diff
@@ -260,9 +258,12 @@ export function getStyles(props: ICoachmarkStyleProps): ICoachmarkStyles {
           },
         },
       },
+      !isMeasuring &&
```
Comment (ecraig12345) [inline packages/react-next/src/components/Coachmark/Coachmark.styles.ts:261]: Why the change in logic here?

File: packages/react-next/src/components/Coachmark/Coachmark.base.tsx [hunk 8]
```diff
@@ -34,126 +33,391 @@ export const COACHMARK_ATTRIBUTE_NAME = 'data-coachmarkid';
  * An interface for the cached dimensions of entity inner host.
  */
 export interface IEntityRect {
-  width: number;
-  height: number;
+  width?: number;
```
Comment (ecraig12345) [inline packages/react-next/src/components/Coachmark/Coachmark.base.tsx:36]: Why are these optional now?

File: packages/react-next/src/components/Coachmark/Coachmark.base.tsx [hunk 9]
```diff
@@ -34,126 +33,391 @@ export const COACHMARK_ATTRIBUTE_NAME = 'data-coachmarkid';
...
+  height?: number;
 }
 
-export interface ICoachmarkState {
+const DEFAULT_PROPS = {
```
Comment (ecraig12345) [inline packages/react-next/src/components/Coachmark/Coachmark.base.tsx:40]:
```suggestion
const DEFAULT_PROPS: Partial<ICoachmarkProps> = {
```

File: packages/react-next/src/components/Coachmark/Coachmark.types.ts (hunk 5 above)
Comment (MLoughry) [reply packages/react-next/src/components/Coachmark/Coachmark.types.ts:232]: They were removed as either unused or entirely redundant. `isBeaconAnimating` was always `true`, and `isMeasured` was always the inverse of `isMeasuring`.

File: packages/react-next/src/components/Coachmark/Coachmark.styles.ts (hunk 7 above)
Comment (MLoughry) [reply packages/react-next/src/components/Coachmark/Coachmark.styles.ts:261]: I don't remember the precise details (this work was done weeks ago), but IIRC the issue was that the measurement was now being taken such that without this change, it would determine that the size of the container to render into was 0x0, making it effectively invisible.

File: packages/react-next/src/components/Coachmark/Coachmark.base.tsx (hunk 8 above)
Comment (MLoughry) [reply packages/react-next/src/components/Coachmark/Coachmark.base.tsx:36]: I don't recall the precise details. Per the [commit message](https://github.com/MLoughry/office-ui-fabric-react/commit/26478ffa47d9330bf1cd165a3b17b8f3a4e0d2ee), this was to fix a bug (currently still in the OUFR version) where initializing the component with `isCollapsed={false}` would not show the expanded Coachmark.

**Comment by MLoughry** [comment]
@ecraig12345 @xugao Could either of you approve the Screener additions?

**Comment by ecraig12345** [comment]
It looks like someone rejected the "Coachmark: Expanded: Default" for current Coachmark because it's not actually showing the expanded coachmark. And in the CoachmarkNext expanded state, I'm not seeing the beak.

image

**Comment by ecraig12345** [comment]
Actually never mind, I don't see a beak on the example on the website either, so maybe the CoachmarkNext one is okay. But I still think the Coachmark current expanded is wrong.
image

**Comment by MLoughry** [comment]
@ecraig12345 I [replied](https://github.com/microsoft/fluentui/pull/13626#issuecomment-658305557) to @xugao about that very issue:
> Yes, there's a logic bug in the original OUFR control when you provide isCollapsed={false} to start with. I fixed the bug when I migrated the control.

**Comment by ecraig12345** [comment]
Oh sorry, I missed that. Maybe we should remove the screener state for now so we're not checking in an incorrect state?

**Comment by ecraig12345** [comment]
I accepted the CoachmarkNext expanded state.

File: packages/react-next/src/components/Coachmark/PositioningContainer/PositioningContainer.tsx [hunk 10]
```diff
@@ -42,397 +42,391 @@ const SLIDE_ANIMATIONS = {
...
+const DEFAULT_PROPS = {
+  preventDismissOnScroll: false,
+  offsetFromTarget: 0,
+  minPagePadding: 8,
+  directionalHint: DirectionalHint.bottomAutoEdge,
+};
 
-  // @TODO rename to reflect the name of this class
-  private _contentHost = React.createRef<HTMLDivElement>();
+function useTargets({ target }: IPositioningContainerProps, positionedHost: React.RefObject<HTMLDivElement | null>) {
```
Comment (ecraig12345) [inline packages/react-next/src/components/Coachmark/PositioningContainer/PositioningContainer.tsx:52]: This is optional, but since PositioningContainer is a fork of Callout, would it be possible to reuse any of the hooks from Callout here?

Comment (MLoughry) [reply packages/react-next/src/components/Coachmark/PositioningContainer/PositioningContainer.tsx:52]: @ecraig12345 I'll do this in a PR fixing ContextualMenu's UNSAFE methods. Since this change has a `package.json` change that causes conflicts after every publish, I'd like to merge this today.

Comment (ecraig12345) [reply packages/react-next/src/components/Coachmark/PositioningContainer/PositioningContainer.tsx:52]: Sounds good.
